### PR TITLE
Update to the latest stable versions of golang, docker-machine,

### PR DESCRIPTION
**Pull request by davidewatson**
terraform, Google Cloud SDK, and kubectl. Use pip (and not apt) to
install all python modules, and use bundler to install all ruby gems,
while fixing the versions installed to those currently tested.
Fixes KRAK-402.

### CODE REVIEW

**Comment by davidewatson** [comment]
Please do not merge this yet. I would like to let it bake on testlet due to the number of version changes.

File: containers/jenkins/Dockerfile [hunk 1]
```diff
@@ -1,27 +1,32 @@
...
-# versions
-# TODO: install from source instead?
-ENV GOLANG_DEB_VERSION 2:1.5.1-4~bpo8+1
-ENV GOOGLE_CLOUD_SDK_VERSION 106.0.0
-ENV KUBERNETES_VERSION v1.2.0
-ENV TERRAFORM_VERSION 0.6.14
-ENV TERRAFORM_EXECUTE_VERSION v0.0.2
-ENV TERRAFORM_COREOSBOX_VERSION v0.0.1
+RUN chmod 400 /root/.ssh/id_rsa \
+  && ssh-keyscan -H github.com >> ~/.ssh/known_hosts
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: IMO these commands are unrelated and shouldn't be chained together for the sake of fewer layers

File: containers/jenkins/Dockerfile [hunk 2]
```diff
@@ -1,27 +1,32 @@
...
 
-# install jenkins plugins
+# Install Jenkins plugins
+COPY plugins.txt /usr/share/jenkins/plugins.txt
 RUN /usr/local/bin/plugins.sh /usr/share/jenkins/plugins.txt
 
-# skip installing gem documentation
+# Configure Ansible
+COPY ansible.cfg /etc/ansible/ansible.cfg
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: maybe configure after it's been installed via pip below?

File: containers/jenkins/Dockerfile [hunk 3]
```diff
@@ -46,26 +48,22 @@ RUN apt-get update \
       rsync \
     && rm -rf /var/lib/apt/lists/*
 
-ENV GOROOT=/usr/lib/go
-RUN mkdir /etc/ansible/
-RUN pip install --upgrade pip
-RUN pip install ansible
-COPY ansible.cfg /etc/ansible/ansible.cfg
+# All python modules should be installed here if possible
+COPY requirements.txt /opt/kraken-ci/containers/jenkins/
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: I dunno how I feel about this container having its name/identity hardcoded into a path.  If we're trying to just use the files, I'd advocate using `/tmp` and cleaning them up.  If you want to leave them around for debugging/forensics, maybe a path like `/opt/build` or something

File: containers/jenkins/Dockerfile [hunk 4]
```diff
@@ -46,26 +48,22 @@ RUN apt-get update \
...
+RUN pip install --upgrade pip \
+  && pip install --requirement /opt/kraken-ci/containers/jenkins/requirements.txt
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: ditto here

File: containers/jenkins/Dockerfile [hunk 5]
```diff
@@ -46,26 +48,22 @@ RUN apt-get update \
...
 
-# install updated jenkins job builder
-RUN pip install PyYAML 
-RUN pip install python-jenkins
-RUN git clone https://git.openstack.org/openstack-infra/jenkins-job-builder /etc/jenkins-job-builder && \
-  cd /etc/jenkins-job-builder && \
-  python setup.py install
+# All ruby gems should be installed here if possible
+ENV BUNDLE_GEMFILE=/opt/kraken-ci/containers/jenkins/Gemfile
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: and here

File: containers/jenkins/Dockerfile [hunk 6]
```diff
@@ -46,26 +48,22 @@ RUN apt-get update \
...
+COPY Gemfile Gemfile.lock /opt/kraken-ci/containers/jenkins/
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: and here

File: containers/jenkins/Dockerfile [hunk 7]
```diff
@@ -46,26 +48,22 @@ RUN apt-get update \
...
+RUN gem install bundler \
+  && bundler install
 
-# install gems
-RUN gem install \
-  bundler:1.11.2 \
-  aruba:0.6.2 \
-  cucumber:2.0.2
+# Install docker machine
+RUN curl -L https://github.com/docker/machine/releases/download/${DOCKER_MACHINE_VERSION}/docker-machine-`uname -s`-`uname -m` > /usr/local/bin/docker-machine \
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: Prefer `$(...)` to backticks

File: containers/jenkins/Dockerfile (hunk 3 above)
Comment (davidewatson) [inline containers/jenkins/Dockerfile]: /opt/kraken-ci is meant to mirror /opt/kraken. I think making the container path the same as the corresponding package path makes the relationship between the files clearer.

File: containers/jenkins/Dockerfile (hunk 7 above)
Comment (davidewatson) [inline containers/jenkins/Dockerfile]: Agreed. Fixed.

File: containers/jenkins/Gemfile [hunk 8]
```diff
@@ -0,0 +1,4 @@
+source "https://rubygems.org"
+
+gem "aruba", "0.6.2"
+gem "cucumber", "~> 2.0.2"
```
Comment (spiffxp) [inline containers/jenkins/Gemfile]: let's prefer pinning directly vs. allowing 2.0.3 to sneak in on us next time we rebuild

File: containers/jenkins/Dockerfile (hunk 2 above)
Comment (davidewatson) [inline containers/jenkins/Dockerfile]: Yes, that is probably safer. Just in case the ansible install overwrites our config... Fixed.

**Comment by spiffxp** [comment]
A few style nits to address, other than that lgtm

File: containers/jenkins/Gemfile (hunk 8 above)
Comment (davidewatson) [inline containers/jenkins/Gemfile]: Fair enough. I verified that cucumber is supposed to obey SemVer 2, but that is of course no guarantee. Fixing.

File: containers/jenkins/Dockerfile (hunk 2 above)
Comment (davidewatson) [inline containers/jenkins/Dockerfile]: I think I am going to move all of the configuration (eg. ssh, jjb) until after installation for consistency.

File: containers/jenkins/Dockerfile (hunk 1 above)
Comment (davidewatson) [inline containers/jenkins/Dockerfile]: I think they are related insofar as they are ssh configuration. In fact, the ssh-keyscan used to be at the end of the Dockerfile, which made the use of tilde in the path unclear. I'll split them and replace the tilde with an absolute path for clarity.

File: containers/jenkins/Dockerfile [hunk 9]
```diff
@@ -46,26 +47,22 @@ RUN apt-get update \
...
-# install updated jenkins job builder
-RUN pip install PyYAML 
-RUN pip install python-jenkins
-RUN git clone https://git.openstack.org/openstack-infra/jenkins-job-builder /etc/jenkins-job-builder && \
-  cd /etc/jenkins-job-builder && \
-  python setup.py install
+# All ruby gems should be installed here if possible
+ENV BUNDLE_GEMFILE=/opt/kraken-ci/containers/jenkins/Gemfile
+COPY Gemfile Gemfile.lock /opt/kraken-ci/containers/jenkins/
+RUN gem install bundler \
```
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: sorry, didn't notice this earlier... can we not pollute the ENV, and instead use `--gemfile=`?  any other invocations of bundler in this container are going to use this Gemfile unless they explicit unset this

File: containers/jenkins/Dockerfile (hunk 3 above)
Comment (spiffxp) [inline containers/jenkins/Dockerfile]: semantics... kraken is dealing with a single container, this is dealing with multiples... IMO I'd find it an easier pattern to go to a known build dir (same location) in any given container, so I can see the Dockerfiles and other forensic artifacts... but let's agree to disagree and find out if sticking the container in the name is easier

File: containers/jenkins/Dockerfile (hunk 9 above)
Comment (davidewatson) [inline containers/jenkins/Dockerfile]: Good idea. Done.

**Comment by spiffxp** [comment]
ok to merge once Gemfile is addressed
